Http retry

arnevogt opened the pull request:
add (configurable) retries for WPS Process

arnevogt committed:
retry if WPS request fails

arnevogt committed:
add wps retries to configuration

arnevogt committed:
edit loggging

nbrinckm commented on asyncwrapper/src/main/kotlin/org/n/riesgos/asyncwrapper/process/wps/WPSProcess.kt:
```diff
@@ -80,66 +84,104 @@ class WPSProcess(private val wpsClient : WPSClientSession, private val url: Stri
+                    LOGGER.info("execute WPS process ${this.processID} (attempt: ${this.retryAttempts})")
+                    val wpsOutput = wpsClient.execute(url, executeRequest, wpsVersion)
+                    processOutput = parseProcessOutput(wpsOutput) //attempt successful
+                    LOGGER.info("executed WPS process ${this.processID} (needed ${this.retryAttempts} attempt(s))")
```
I guess I need to do the review tomorrow in more detail, but would it make sense to reset the retry count here?

nbrinckm commented on asyncwrapper/src/main/kotlin/org/n/riesgos/asyncwrapper/process/wps/WPSProcess.kt:
```diff
@@ -1,19 +1,23 @@
-class WPSProcess(private val wpsClient : WPSClientSession, private val url: String, private val processID: String, private val wpsVersion: String, private val expectedOutputs : List<WPSOutputDefinition>) : Process {
+class WPSProcess(private val wpsClient : WPSClientSession, private val url: String, private val processID: String, private val wpsVersion: String, private val expectedOutputs : List<WPSOutputDefinition>, private val retryConfig : RetryConfiguration) : Process {
+
+    private var retryAttempts = 0;
```
As far as possible, I would not make the retry count a local variable. Instead I would prefer to have it local in the methods themselves, so we don't have to care about its current state in the WPSProcess instance itself.

nbrinckm commented on asyncwrapper/src/main/kotlin/org/n/riesgos/asyncwrapper/process/wps/WPSProcess.kt:
```diff
@@ -1,19 +1,23 @@
+class WPSProcess(private val wpsClient : WPSClientSession, private val url: String, private val processID: String, private val wpsVersion: String, private val expectedOutputs : List<WPSOutputDefinition>, private val retryConfig : RetryConfiguration) : Process {
+
+    private var retryAttempts = 0;
+    constructor(wpsClient : WPSClientSession, url: String, processID: String, wpsVersion: String, expectedOutputs : List<WPSOutputDefinition>) : this(wpsClient, url, processID, wpsVersion, expectedOutputs, RetryConfiguration(3, 3000))
```
With the overall config in the main constructor, do we still need this side constructor?

nbrinckm replied:
Still a question: Do we use this secondary constructor? If we need it, then we don't give the arguments from the application.yml - which is not what we want.

I would like you to remove that secondary constructor.

nbrinckm commented on asyncwrapper/src/main/kotlin/org/n/riesgos/asyncwrapper/process/wps/WPSProcess.kt:
```diff
@@ -80,66 +84,104 @@ class WPSProcess(private val wpsClient : WPSClientSession, private val url: Stri
-                        LOGGER.info("Stored result for " + complexOutput.id)
+            while(this.retryAttempts < this.retryConfig.attempts) {
+                try {
+                    this.retryAttempts++
```
Honestly I would increase the retry attempts only in the code of the error handling. (Basically when I try it the first time, the retry counter should be 0. It should only increase when I got an error & decide that I want to retry).

I guess it is still working to have it here, but it feels a bit odd to me. (But I would also let myself be outvoted here)

nbrinckm commented on asyncwrapper/src/main/kotlin/org/n/riesgos/asyncwrapper/process/wps/WPSProcess.kt:
```diff
@@ -80,66 +84,104 @@ class WPSProcess(private val wpsClient : WPSClientSession, private val url: Stri
+                    this.retryAttempts++
+                    LOGGER.info("execute WPS process ${this.processID} (attempt: ${this.retryAttempts})")
+                    val wpsOutput = wpsClient.execute(url, executeRequest, wpsVersion)
+                    processOutput = parseProcessOutput(wpsOutput) //attempt successful
```
Do we have another point where we access the processOutput?
If not we could move the declaration into the while loop (and handle it as `val`).

nbrinckm commented:
I'm not 100% through the code here (I'm going to do that tomorrow), but the question: You did't add retries for establishing the connection to the wps server, did you? 
I found that I needed it in #60. I think it would make sense to adopt it to the configurability as you did it here in your branch.

nbrinckm commented on asyncwrapper/src/main/kotlin/org/n/riesgos/asyncwrapper/config/RetryConfiguration.kt:
```diff
@@ -0,0 +1,7 @@
+import org.springframework.boot.context.properties.ConstructorBinding
+
+
+data class RetryConfiguration(var attempts: Int = 3, var backoff_millis: Long = 3000)
```
Snake case in kotlin/java looks kind of strange to me. Can we rename it to `backoffMillis` (or even `sleeptimeMsBetweenRetries`)?

nbrinckm replied:
Also the question: Do we need to have those fields modifiable (`var`)? I would favor immutability (`val`) whereever this is possible.

nbrinckm commented:
One thing that came into my mind yesterday: Can we solve the retry thing with a decorator pattern?

So that we can write something like this:

```
val processOutput = doAndRetryIfNeeded<WPSClientException>(retryConfig) { retryCount ->
    LOGGER.info("bla with ${retryCount}")
    wpsOutput = wpsClient.execute()
}
```

I think a decorator as this would be reusable for many different things - but I'm not sure if the implementation would be trivial. I think it would need some kind of static function, so that we can keep the `this` context of the class. I haven't tried it yet. But maybe this is something where it really makes sense to have kotlin & its power regarding higher order functions.

Commit:
First version of a generic retry decorator

arnevogt committed:
merge main

arnevogt committed:
Merge remote-tracking branch 'origin/retry-decorator' into httpRetry

arnevogt committed:
use RetryUtils for WPS execute

arnevogt committed:
change retry semantics

arnevogt committed:
remove redundant instance variable

arnevogt committed:
implement retries for retrieval of wps process description

arnevogt committed:
use RetryUtils for WPS getCapabilities request

arnevogt commented:
@nbrinckm 

- adopted RetryUtils decorator for getCapabilites, getProcessDescription and executeProcess
- change retry semantics (retries do not include first attempt)

nbrinckm commented on asyncwrapper/src/main/kotlin/org/n/riesgos/asyncwrapper/utils/RetryUtils.kt:
```diff
@@ -0,0 +1,35 @@
+            // ignore this specific exception and retry
+            retryCount++
+            lastException = ex
+            Thread.sleep(sleepTimeMilliSeconds)
```
I realized that we also sleep if we failed on the last attempt - one of the errors I made when I wrote the first version of this retry decorator.

Maybe we can a simple if statement here, so that we don't sleep when we already know that there will be no last attempt.

nbrinckm commented:
Maybe we can add the retry also in the fetchContent method in the AbstractWrapper file. 

Here we fetch the contents from the object storage - and http retry settings could be helpful as well. When I take a look at https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpClient.html then it seems that this could throw IOExceptions or InterruptedException. I guess the IOException is the better one to catch here.

Same is true for the file storage upload call (currently line 435 in AbstractWrapper.kt). Here it makes sense to retry the upload if it fails due to network reasons. The exceptions here are quite a lot, however I added the wrote the upload method on my own in Java (due to a conflict with the `object` method name in the MinIOClient & the keyword in kotlin). We could just write our own Exception class that we can fill with any of the problematic exceptions from the MinIOClient & just catch our own exception in the retry.

We could also change the decorator to give it a predicate function that tests the exception. That way we could handle multiple exception types.

Could look like this:

```
val wpsProcess = retry<WPSProcess>(
  forExceptionsOfTypes(
     IOException::class.java., InterruptedException::class.java
  ), { it ->
     return@retry WPSProcess()
})
```

I think in that case we could even remove the exception type parameter from that function & just catch every possible exception - that we then put into the predicate.



But maybe all of those ideas go out of scope of that MR here, so feel free to make a separate issue if you think we can handle that better externally.

nbrinckm commented:
My base points are the secondary constructor & maybe the improvements of the very last sleep call. Then I will approve the pull request. Rest is optional & can be made in other issues &  pull requests.

MichaelLangbein commented:
> But maybe all of those ideas go out of scope of that MR here, so feel free to make a separate issue if you think we can handle that better externally.

Hi Nils! 
As for my part, I think indeed that the improvements you mentioned are better dealt with in a separate issue - just because I don't want you to have to keep working on this after the sprint is officially over.
I've created one here: https://github.com/riesgos/async/issues/68
And here: https://github.com/riesgos/async/issues/69

arnevogt committed:
do not sleep if final retry failed

arnevogt committed:
remove redundant secondary constructor for WPSProcess

arnevogt commented:
@nbrinckm 

pls re-review

- removed unnecessary sleep after final retry failed
- removed secondary constructor for WPSProcess
  - actually use retry config

arnevogt commented:
I will address the other improvements in another PR.

nbrinckm reviewed (approved):
Looks good from my point of view.
Feel free to merge.
